### ISSUE TITLE
Support for Set and Map

### ISSUE THREAD

**Issue by cmmartin** [immerjs/immer#146]
* [x] Feature request: Is it possible to support mutating the native `Set`, `Map`, `WeakMap`, etc?
  * [ ] I am willing to implement this myself

**Comment by RikuVan** [comment]
I had a quick try at adding Set and Map for es6 environment. I don't completely get everything happening in immer, so I well may have missed some complexities. It at first glance it seems actually pretty easy to proxy these since all mutations happen through method calls: 
```js
const nonMutatingMethods = [
    "forEach",
    "get",
    "has",
    "keys",
    "set",
    "values",
    "entries",
    "@@iterator"
]

const mutatingMethods = ["set", "delete", "clear", "add"]

const properties = ["size"]

function markMapOrSetChanged(proxy) {
    proxy.modified = true
    proxy.copy = shallowCopy(proxy.base)  // e.g. new Map(proxy.base)
    if (proxy.parent) markChanged(proxy.parent)
    nonMutatingMethods.forEach(function(method) {
        proxy[method] = function(a, b) {
            return proxy.copy[method](a, b)
        }
    })
    properties.forEach(function(prop) {
        proxy[prop] = proxy.copy[prop]
    })
    return proxy
}

function proxyMapOrSet(state) {
    var proxyState = {}
    var proxy = state
    nonMutatingMethods.forEach(function(method) {
        proxy[method] = function(a, b) {
            return proxy.base[method](a, b)
        }
    })
    mutatingMethods.forEach(function(method) {
        proxy[method] = function(key, value) {
            if (!proxy.modified) markMapOrSetChanged(proxy)
            proxy.copy[method](key, value)
        }
    })
    Object.defineProperty(proxy, PROXY_STATE, {
        value: proxy,
        enumerable: false,
        writable: true
    })
    proxyState.proxy = proxy
    return proxyState
}
```
I guess you don't need to recurse into the Map/Set values since they are all set through method calls?

**Comment by RikuVan** [comment]
but it seems you can't freeze these, which means you have to do something kind of hacky there I guess:
```
function freezeMapOrSet(value) {
    const insertMethod = isMap(value) ? "set" : "add"
    if (autoFreeze) {
        ;[insertMethod, "delete", "clear"].forEach(function(method) {
            Object.defineProperty(value, method, {
                value: function() {
                    throw "In a frozen state, cannot be mutated"
                },
                enumerable: false,
                writable: false
            })
        })
        Object.freeze(value)
    }
    return value
}
```

**Comment by bradennapier** [comment]
So I wanted to get full support for ES6 Map/Set.  I also needed to get change logs of every update -- since the proxy makes this pretty efficient I took my own stab at re-building the library a bit.  

## [immuta](https://github.com/odo-network/immuta)

It is still a little messy since I was just getting it into the fold but the example does a good example of showing it in action:

```javascript
import immuta from "immuta";
import printDifference from "immuta/utils/print-difference";

const state = {
  deep: {
    foo: {
      bar: {
        baz: true
      }
    },
    set: new Set([{ one: "two" }, { two: "three" }]),
    map: new Map([["one", { foo: "bar" }]]),
    array: [{ i: 1 }, { i: 2 }, { i: 3 }, { i: 4 }, { i: 5 }]
  }
};

const next = immuta(
  // provide the state to start with
  state,
  // draft is a proxy that will copy-on-write
  draft => {
    const one = draft.deep.map.get("one");
    if (one) {
      one.foo = 1;
    }
    draft.deep.set.clear();
    draft.deep.set.add({ some: "obj" });

    draft.deep.array[2].foo = "bar!";
  },
  // optional callback for change events
  (changedState, changedMap, rollback) => {
    // rollback() will cancel the changes and return original object
    // changedMap is Map { 'path.to.change' => changedValue }
    // changedState is the new state being returned to caller (nextState)
    changedMap.forEach((v, changedKey) => {
      console.log("Change: ", changedKey);
    });
  }
);

printDifference(state, next);
```

```
Change:  deep
Change:  deep.map(=> [MAP])
Change:  deep.map(=> [MAP]).get.(one)
Change:  deep.map(=> [MAP]).get.(one).foo
Change:  deep.set(=> [SET])
Change:  deep.array
Change:  deep.array.2
Change:  deep.array.2.foo
```

![image](https://user-images.githubusercontent.com/15365418/40952263-6d416650-682f-11e8-95bc-4e247ec48098.png)

**Comment by RikuVan** [comment]
Cool, I am curious to look at your version more carefully as soon as I have more time.

**Comment by bradennapier** [comment]
Yeah it is working well now and performs very well.  As for mentioning not needing to recurse, I disagree.  Take this example:

```javascript
const state = {
  deep: {
    foo: {
      bar: {
        baz: true,
      },
    },
    set: new Set([{ one: 'two' }, { two: 'three' }]),
    map: new Map([['one', { foo: 'bar' }]]),
    array: [{ i: 1 }, { i: 2 }, { i: 3 }, { i: 4 }, { i: 5 }],
  },
};

const next = immuta(
  // provide the state to start with
  state,
  // draft is a proxy that will copy-on-write
  draft => {
    const val = draft.deep.map.get('one');
    foo.baz = 'qux';
  },
  // optional callback for change events
  (changedState, changedMap, rollback) => {
    // rollback() will cancel the changes and return original object
    // changedMap is Map { 'path.to.change' => changedValue }
    // changedState is the new state being returned to caller (nextState)
    changedMap.forEach((v, changedKey) => {
      console.log('Change: ', changedKey);
    });
  },
);
```

**Comment by bradennapier** [comment]
I provided some more insight into some of the challenges with `Set` (and a little bit into `Map`) with the release of some extra helpers to handle deep merging.  They may also be helpful to anyone who might want to take on such a task.  

https://github.com/odo-network/immuta/issues/1

Essentially `Set` introduces a fairly difficult problem in that it when you want to change the reference to the item in the `Set` you actually need to delete the original value which ends up changing the insertion order of the Set. 

Merge is essentially impossible because of this, when insertion order is no longer maintained there is really no way to "map" the desired values to their originals that we want to merge with.

**Comment by mweststrate** [comment]
For now added some docs on how to work with Map's and Set's in combination with Immer:

https://github.com/mweststrate/immer/commit/d9b4d566f4e6043384ab10a758c5ff943b33f57a

Closing this issue for now, but if  PR shows up that supports both Maps and Sets completely, with deep updates and also works in ES5 mode :)

**Comment by szagi3891** [comment]
Someone might want to add this PR?  :)

Support for maps and set is the last thing that is missing in this cool library. Then I could get rid of the dependence on immutable.js

**Comment by keenondrums** [comment]
@mweststrate what if a PR for ES6 only is going to show up? Moreover, what if it is only for Maps?
Let me be honest, I use Maps a lot and I really really would like immer to have fo have first-class support for them. Yet, I, fortunately, do not have to support ES5 envs in my projects. So I'm wondering if it makes sense for me to work on a PR for immer which includes at least (but not at most) ES6 support for Maps only?
My priorities would be:
1. (Required minimum) ES6 Maps support 
1. (Desireable, but not guaranteed) ES6 Sets support
1. (Optional, would work on it only if it doesn't take too much of my time) ES5 Maps support
1.  (Optional, would work on it only if it doesn't take too much of my time)ES5 Sets support

Would it be beneficial for the community to have at least ES6 Maps support? Would you be willing to merge such a PR?

**Comment by keenondrums** [comment]
@mweststrate I also have a question on the design. Would you expect a mutation of a Map item to cause a creation of a new Map?

In other words, should this work?
```js
const state = {
   users: new Map([["michel", {name: "miche"}]])
}

const nextState = produce(state, draft => {
   const user = draft.users.get("michel")
   // Should this one work?
   user.name = 'Aladdin'
})
```

**Comment by mweststrate** [comment]
Yes :)

**Comment by keenondrums** [comment]
So far I managed to make an early preview of nested updates for ES6 Maps. The ongoing work can be tracked [here](https://github.com/keenondrums/immer/pull/1/files). If anyone is interested in providing any kind of help with this feature please email me at andrey.goncharov+it@protonmail.com

**Comment by keenondrums** [comment]
@mweststrate PR for ES6 maps https://github.com/immerjs/immer/pull/350

**Comment by runnez** [comment]
@mweststrate @aleclarson Guys look at this https://github.com/immerjs/immer/pull/437

**Comment by mweststrate** [comment]
If the PR is solid, it's coming, but didn't have any time to review yet.

**Comment by vytautas-pranskunas-** [comment]
It would really help us because we have built our state with Set and
appears that it became mutable and mutates primary object...

**Comment by mweststrate** [comment]
This has been released as Immer 5.0!
